timesource/timesource_test.go

```diff
@@ -0,0 +1,113 @@
+package timesource
+
+import (
+	"errors"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/beevik/ntp"
+	"github.com/stretchr/testify/assert"
+)
+
+const (
+	// clockCompareDelta declares time required between multiple calls to time.Now
+	clockCompareDelta = 30 * time.Microsecond
+)
+
+type testCase struct {
+	description string
+	attempts    int
+	responses   []queryResponse
+	expected    time.Duration
+	expectError bool
+
+	// actual attempts are mutable
+	mu             sync.Mutex
+	actualAttempts int
+}
+
+func (tc *testCase) query(string) (*ntp.Response, error) {
+	tc.mu.Lock()
+	defer func() {
+		tc.actualAttempts++
+		tc.mu.Unlock()
+	}()
+	response := &ntp.Response{ClockOffset: tc.responses[tc.actualAttempts].Offset}
+	return response, tc.responses[tc.actualAttempts].Error
+}
+
+func newTestCases() []*testCase {
+	return []*testCase{
+		{
+			description: "SameResponse",
+			attempts:    3,
+			responses: []queryResponse{
+				{Offset: 10 * time.Second},
+				{Offset: 10 * time.Second},
+				{Offset: 10 * time.Second},
+			},
+			expected: 10 * time.Second,
+		},
+		{
+			description: "Median",
+			attempts:    3,
+			responses: []queryResponse{
+				{Offset: 10 * time.Second},
+				{Offset: 20 * time.Second},
+				{Offset: 30 * time.Second},
+			},
+			expected: 20 * time.Second,
+		},
+		{
+			description: "Error",
```

Comment:
Even if one of N attempts fails, we discard all results? Is this expected?

Reply:
yes, i thought to change it so that 5 requests are made and 2 failures can be tolerated, but then decided that it is not worth it